### PR TITLE
feat: teach rulesets to accept exceptions

### LINKED ISSUE #747: Rule exceptions
**User story.**
As a user, I can add exceptions to certain rules to ignore failures so that I can guard against new changes but run against legacy descriptions.

**Is your feature request related to a problem?**

For API descriptions that have a "pre-linting" era, but with linting rules that should be enforced for new changes, it would be nice to have some kind of mechanism similar to eslint's `// eslint-disable-line no-use-before-define`.

**Describe the solution you'd like**

I couldn't find any talks about this feature in issues / PRs, let me know if it exists already. I can see different ways of tackling this.

  - YAGNI: Users are expected to run spectral in a custom way, running only against new changes / diffs. Or maybe expected to write custom JS rules and do the filtering there.

  - Users are expected to handle this with `given`'s JSON PATH entirely.

  - A new rule attribute:

```
  paths-kebab-case:
    description: Should paths be kebab-case.
    message: '{{property}} is not kebab-case: {{error}}'
    severity: error
    recommended: true
    given: $.paths[*]~
    then:
      function: pattern
      functionOptions:
        match: "^(\/[a-z0-9-{}]+)+$"
  except:
    - <json_path_matching-legacy-paths>
```

  - An eslint style convention directly on API specifications (Tricky since users would be required to strip this out before making public)

```
x-spectral-disable: "paths-kebab-case"
```

### PR DESCRIPTION
**Pull request by nulltoken**
Fixes #747.

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated
- [x] Add a STDIN test
- [ ] Create an issue to explicit the potential benefit of not swallowing up the ignored results
- [x] Create an issue to keep track of STDIN as a potential source (#1001)
- [x] Create an issue to keep track of potential changes to the exceptions engine to make it laxer (#1002)

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/rulesets/reader.ts [hunk 1]
```diff
@@ -97,6 +101,7 @@ const createRulesetProcessor = (
         if (extendedRuleset !== null) {
           mergeRules(rules, extendedRuleset.rules, parentSeverity);
           Object.assign(functions, extendedRuleset.functions);
+          mergeExceptions(baseUri, exceptions, extendedRuleset.exceptions);
```
Comment (P0lip) [inline src/rulesets/reader.ts]: Food for thoughts, but I'm pretty sure we'll want to filter out certain exceptions.
For instance,
```json
{
  "file/#: ["foo-rule"],
  "file/#.bar": ["foo-rule"]
}
```
This could be expressed in a single exception `"file/#": ["foo-rule"]`.
We'll have fewer paths to query later if we narrow them down here.

**Comment by m-mohr** [comment]
I like this a lot!

**Comment by nulltoken** [comment]
@philsturgeon @P0lip Here's a first alpha version of the PR.

***repro/one.yaml***
```
openapi: 3.0.0
info:
  version: 1.0.0
  title: Stoplight
paths: {}
```

***repro/ruleset.yaml***
```
extends: spectral:oas

except:
  "one.yaml#":
    - oas3-api-servers
  "./one.yaml#/info":  # Beside relative ones, also supports rooted paths ;-)
    - info-contact
```
***cli usage***
```
$ yarn cli lint repro/one.yaml -r .epro/ruleset.yaml
OpenAPI 3.x detected

f:/spectral/repro/one.yaml
 1:1  warning  openapi-tags      OpenAPI object should have non-empty `tags` array.
 2:6  warning  info-description  OpenAPI object info `description` must be present and non-empty string.

✖ 2 problems (0 errors, 2 warnings, 0 infos, 0 hints)
```

I'm going to start working on the doco.
Still wrestling a bit with a harness related issue, but should be ready for a first review pass.

/cc @m-mohr Would you agree to give it a try against your own use cases, I'd be delighted to get some feedback!

File: src/utils/sortResults.ts [hunk 2]
```diff
@@ -44,23 +45,29 @@ const normalize = (value: number): -1 | 0 | 1 => {
   return 0;
 };
 
+export const comparePosition = (left: IPosition, right: IPosition): -1 | 0 | 1 => {
```
Comment (nulltoken) [inline src/utils/sortResults.ts]: This is actually unrelated to the gist of this PR. I initially started with a range based comparison approach. That proved to be a bad idea. However, while building the now defunct `rangeEquals`, I refactored that piece of code.

Although quite a minor change, I decided to keep it along (as it cleans the code a little) and covered it with tests.

File: src/__tests__/linter.jest.test.ts [hunk 3]
```diff
@@ -170,4 +174,176 @@ describe('Linter', () => {
...
+
+      expect(res).not.toContainEqual(
+        expect.objectContaining({
+          code: 'oas3-api-servers',
+        }),
+      );
+    });
+
+    describe('resolving', () => {
+      const testRules: Dictionary<Rule, string> = {
```
Comment (nulltoken) [inline src/__tests__/linter.jest.test.ts]: @P0lip I now realize this whole pre-test setup is a lot of code. Do you want me to try and repackage it, or is it good enough?

File: test-harness/index.ts [hunk 4]
```diff
@@ -26,17 +26,21 @@ describe('cli acceptance tests', () => {
...
-          );
+          const normalizedName = normalize(tmpFileHandle.name);
+
+          replacements[asset] = normalizedName;
+          replacements[`${asset}|no-ext`] = normalizedName.replace(new RegExp(`${path.extname(normalizedName)}$`), '');
+        }),
+      );
 
-          await writeFileAsync(tmpFileHandle.name, contents, { encoding: 'utf8' }); // todo: apply replacements to contents
+      await Promise.all(
```
Comment (nulltoken) [inline test-harness/index.ts]: @P0lip Splitting this in two parts protects us from race conditions (eg. before replacing the name of the document in the ruleset, we need to be sure the document tempory storage file has been created and its location has been inserted in the replacements dictionary)

**Comment by nulltoken** [comment]
@P0lip As this builds upon #966, maybe reviewing first #966 (and maybe (with a little luck 🤞 ) getting it merged 😉 ) would reduce the diff noise in this one and make slightly easier to review.

File: src/utils/sortResults.ts (hunk 2 above)
Comment (nulltoken) [reply src/utils/sortResults.ts]: @P0lip I eventually dropped it from here and promoted it into its own PR (#967)

File: src/runner.ts [hunk 5]
```diff
@@ -4,17 +4,36 @@ import { DocumentInventory } from './documentInventory';
...
   documentInventory: DocumentInventory,
   rules: RunRuleCollection,
   functions: FunctionCollection,
+  exceptions: RulesetExceptionCollection,
 ): IRuleResult[] => {
   const results: IRuleResult[] = [];
 
+  // TODO:
+  // Ruleset
+  // - To be investigated: How to cope with STDIN as a source and exception relative locations?
```
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/runner.ts]: @P0lip As I'm not sure how to build a ruleset that would be able to reference a document passed through STDIN, I'm wondering if issuing a warning wouldn't be the best short-term approach.

When the ruleset contains exceptions
And the document is passed as STDIN
A warning would be issued notifying that exceptions handling isn't supported in that mode and that the run will ignore them.

Thoughts?

File: src/rulesets/reader.ts (hunk 1 above)
Comment (nulltoken) [reply src/rulesets/reader.ts]: @P0lip Is that a use case so common that it should be included in that PR or should we wait until we get some feedback that maintaining exceptions is too tedious?

**Comment by danielgtaylor** [comment]
I like this approach, but what do you think of also supporting something in the OpenAPI itself, e.g. something like this:

```yaml
openapi: 3.0.0
info:
  version: 1.0.0
  title: Stoplight
  x-linter-ignore:
    - info-contact
x-linter-ignore:
  - oas3-api-servers
paths: {}
```

Then the exceptions can be explicitly listed in the API description document rather than having to link two places via paths in separate documents, and it prevents the need for an almost empty `.spectral.yaml` when using a shared set of rules.

You'd have to decide if the rules apply to all children as well, e.g. if ignoring a camel-casing rule on a path would apply to all parameters for all operations under that path.

File: src/runner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (philsturgeon) [reply src/runner.ts]: If you STDIN a file, it may well still have $ref's, and those $ref's would still have exceptions working just the same right? It'd just be the initial "entry file" being passed in as STDIN, and that could have things ignored with `{ENTRY FILE OR ROOT IDENTIFIER}#/components` and not mentioning a filename? Or do we need some sort of wildcard for saying "this is the first/main/root/entry file?

**Comment by nulltoken** [comment]
@danielgtaylor Indeed. Your proposal reminds me of one that was described in the initial issue #747.

File: src/runner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/runner.ts]: @philsturgeon I'm not even sure how a $ref containing a relative file  location would be resolved in that context and I can't find any test in the codebase about that use case. Does it consider the current working directory as the root to be used?

@P0lip Any idea?

File: test-harness/helpers.ts [hunk 6]
```diff
@@ -99,16 +99,17 @@ export function tmpFile(opts?: tmp.TmpNameOptions): Promise<tmp.FileResult> {
...
-  let result: RegExpExecArray | null;
-
-  // tslint:disable-next-line:no-conditional-assignment
-  while ((result = BRACES.exec(str))) {
-    if (!(result[1] in values)) continue;
-    const newValue = String(values[result[1]] || '');
-    str = `${str.slice(0, result.index)}${newValue}${str.slice(BRACES.lastIndex)}`;
-    BRACES.lastIndex = result.index + newValue.length;
-  }
+  const replacer = (match: string, identifier: string): string => {
```
Comment (P0lip) [inline test-harness/helpers.ts]: @nulltoken could you rebase the changes from develop?

File: docs/getting-started/rulesets.md [hunk 7]
```diff
@@ -64,6 +64,47 @@ rules:
 In most cases, you will want to operate on resolved document and therefore won't specify that property.
 You might find `resolved` useful if your rule requires access to $refs.
 
+### Given
```
Comment (P0lip) [inline docs/getting-started/rulesets.md]: Nice, appreciate it!
Could we actually create a separate PR out of it so that it makes its way as a part of 5.0.1? Will be useful from the online docs perspective

File: src/__tests__/linter.jest.test.ts (hunk 3 above)
Comment (P0lip) [reply src/__tests__/linter.jest.test.ts]: Yeah, I believe we could create a simple ruleset a put in a different file and simply load it here via `loadRuleset`? What do you think?

File: src/linter.ts [hunk 8]
```diff
@@ -76,5 +78,32 @@ export const lintNode = (
     );
   }
 
-  return results;
+  const isAKnownException = (violation: IRuleResult, locations: IExceptionLocation[]): boolean => {
```
Comment (P0lip) [inline src/linter.ts]: Could we move this function out of `lintNode`? Doesn't seem like it accesses anything out of its own scope.

File: src/rulesets/__tests__/validation.test.ts [hunk 9]
```diff
@@ -242,6 +242,20 @@ describe('Ruleset Validation', () => {
...
   });
+
+  describe('Exceptions validation', () => {
+    const rulesetsWithInvalidExceptStructures = [
+      { extends: ['foo'], except: '' },
+      { extends: ['foo'], except: { one: null } },
+      { extends: ['foo'], except: { one: [1] } },
+    ];
+
+    it.each(rulesetsWithInvalidExceptStructures)('throws when defined "except" do not match schema', async r => {
```
Comment (P0lip) [inline src/rulesets/__tests__/validation.test.ts]:
```suggestion
    it.each(rulesetsWithInvalidExceptStructures)('throws when defined "except" do not match schema', ruleset => {
```

File: src/rulesets/__tests__/validation.test.ts [hunk 10]
```diff
@@ -242,6 +242,20 @@ describe('Ruleset Validation', () => {
...
+      expect(() => {
+        assertValidRuleset(r);
```
Comment (P0lip) [inline src/rulesets/__tests__/validation.test.ts]:
```suggestion
        assertValidRuleset(ruleset);
```

File: test-harness/index.ts (hunk 4 above)
Comment (P0lip) [reply test-harness/index.ts]: Yeah, sounds good to me.

File: src/spectral.ts [hunk 11]
```diff
@@ -135,6 +137,17 @@ export class Spectral {
     }
   }
 
+  private setExceptions(exceptions: RulesetExceptionCollection) {
+    const target: RulesetExceptionCollection = {};
+    mergeExceptions(target, exceptions);
+
+    empty(this.exceptions);
+
+    Object.entries(target).forEach(([location, rules]) => {
```
Comment (P0lip) [inline src/spectral.ts]: ```Object.assign(this.exception, target)``` should be sufficient I believe? It doesn't create a new array for rules though, but if you do need it, you can just clone the target beforehand.

File: src/runner.ts [hunk 12]
```diff
@@ -4,17 +4,36 @@ import { DocumentInventory } from './documentInventory';
...
+  // LoadRuleSet
+  // - TODO: Find out how paths that escape out of the root behave
+  // Spectral
+  // - idea: report exceptions hit through a new 'Silenced' Severy level (-2) -> This opens up a wide range of possible user friendly reporting
```
[the remark this answers is not in the document]
Comment (P0lip) [reply src/runner.ts]: This wouldn't play well with our ecosystem :/
Could use `hint` or `debug` perhaps?

File: src/runner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (P0lip) [reply src/runner.ts]: IIRC, the document path would be equal to the output of `process.cwd()`.
I wouldn't really bother too much though, as it's kind of expected you provide a dereferenced document as STDIN.

File: src/rulesets/mergers/exceptions.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+  const pointer = extractPointerFromRef($ref);
+
+  if (typeof pointer !== 'string') {
+    throw new Error(buildInvalidUriErrorMessage($ref, rulesetUri, 'Missing pointer fragment'));
+  }
+
+  try {
+    pointerToPath(pointer);
+  } catch {
+    throw new Error(buildInvalidUriErrorMessage($ref, rulesetUri));
```
Comment (P0lip) [inline src/rulesets/mergers/exceptions.ts]: How about having a dedicated error, such as `InvalidUriError`? It'd be somewhat similar to ValidationError we have in case of rulesets.

File: src/rulesets/mergers/exceptions.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+import { RulesetExceptionCollection } from '../../types/ruleset';
+
+const normalize = ($ref: string, rulesetUri?: string): string => {
+  const source = extractSourceFromRef($ref);
+
+  if (typeof source !== 'string') {
+    throw new Error(buildInvalidUriErrorMessage($ref, rulesetUri, 'Missing source'));
+  }
+
+  if (rulesetUri === undefined && !isAbsolute(source)) {
```
Comment (P0lip) [inline src/rulesets/mergers/exceptions.ts]:
```suggestion
  if (rulesetUri === void 0 && !isAbsolute(source)) {
```

File: src/__tests__/spectral.test.ts [hunk 15]
```diff
@@ -1,12 +1,15 @@
...
-import { merge } from 'lodash';
+import { escapeRegExp, merge } from 'lodash';
 
 import { Document } from '../document';
 import * as Parsers from '../parsers';
 import { Spectral } from '../spectral';
 import { IResolver, IRunRule, RuleFunction } from '../types';
+import { RulesetExceptionCollection } from '../types/ruleset';
+
+import { buildRulesetExceptionCollectionFrom } from '../rulesets/mergers/__tests__/exceptions.test';
```
Comment (P0lip) [inline src/__tests__/spectral.test.ts]: TBH, I'm not a fun of referencing other tests in test.

Consider exposing that function in `setupTests.ts` as a global fn.

File: src/utils/pivotExceptions.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+export const pivotExceptions = (
+  exceptions: RulesetExceptionCollection,
+  runRules: Dictionary<IRunRule, string>,
+): Dictionary<IExceptionLocation[], string> => {
+  const dic: Dictionary<IExceptionLocation[], string> = {};
+
+  Object.entries(exceptions).forEach(([location, rules]) => {
+    const pointer = extractPointerFromRef(location)!;
```
Comment (P0lip) [inline src/utils/pivotExceptions.ts]: Should we account for lacking pointers? I'm pretty sure we validate them for the most of the time at the moment, so there won't be a case when there is a local ref provided, but what if that function is used in a different context?

File: test-harness/helpers.ts (hunk 6 above)
Comment (nulltoken) [reply test-harness/helpers.ts]: @P0lip Fixed.

File: src/runner.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/runner.ts]: @P0lip Ok. Another proposal: extend the `IRuleResult` interface to make it expose a `silenced` boolean property. Thoughts?

File: src/__tests__/linter.jest.test.ts (hunk 3 above)
Comment (nulltoken) [reply src/__tests__/linter.jest.test.ts]: @P0lip Fixed

File: src/linter.ts (hunk 8 above)
Comment (nulltoken) [reply src/linter.ts]: @P0lip Fixed

File: src/rulesets/__tests__/validation.test.ts (hunk 9 above)
Comment (nulltoken) [reply src/rulesets/__tests__/validation.test.ts]: @P0lip Fixed

File: src/rulesets/__tests__/validation.test.ts (hunk 10 above)
Comment (nulltoken) [reply src/rulesets/__tests__/validation.test.ts]: @P0lip Fixed

File: src/spectral.ts (hunk 11 above)
Comment (nulltoken) [reply src/spectral.ts]: @P0lip Fixed

File: src/rulesets/mergers/exceptions.ts (hunk 14 above)
Comment (nulltoken) [reply src/rulesets/mergers/exceptions.ts]: @P0lip Fixed

File: src/runner.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (P0lip) [reply src/runner.ts]: > @P0lip Ok. Another proposal: extend the IRuleResult interface to make it expose a silenced boolean property. Thoughts?

I'm not sure. Perhaps we simply don't log such errors at all?
We could expose a flag such as `--ignore-skipped-diagnostics` or something similar that would be true by default, similarly to what we do with unknown formats.
If one wants to have ignored diagnostics, that flag could be passed and set to false.

What do you think?
cc @philsturgeon - what's your take on that?

[the remark this answers is not in the document]
Comment (philsturgeon) [reply src/runner.ts]: We are using many different words here all of which have distinct meaning. 

- Diagnostics - this is a new term, lets avoid introducing it.
- Violations - this term is new to the user too, we've always talked about "results". Violations exist in Prism but they're "Contract Violations", so lets try and avoid this term being advertised too
- Exceptions - usually means `throw new Exception` so this might cause confusion
- Ignore - Even if its run internally, this thing doesn't exist to the user because it is completely ignored
- Skipped - something should have happened but we didnt do it, maybe mention it though to let them know - like a skipped test in Jest

We want to ignore. Ignore a rule, ignore a file, pretend it doesn't exist. No output, no warning, no nothing anywhere.

File: src/runner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (philsturgeon) [reply src/runner.ts]: yeah CWD.

**Comment by philsturgeon** [comment]
@danielgtaylor I'd like to keep the scope of this PR to ruleset based ignore, but we could have another issue discussing the pros and cons of letting API description authors mark their own descriptions with ignores. 

It could be seen as a bit cheeky, trying to squeak an API with "naughty things" past automated API governance, but... make a case for it in a new issue and we'll see! :D

File: src/runner.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/runner.ts]: @philsturgeon I agree. Let's not extend the scope of that PR. Once this is merged, I'll open a dedicated issue on that topic.

File: src/rulesets/mergers/exceptions.ts (hunk 13 above)
Comment (nulltoken) [reply src/rulesets/mergers/exceptions.ts]: @P0lip Fixed

File: src/__tests__/spectral.test.ts (hunk 15 above)
Comment (nulltoken) [reply src/__tests__/spectral.test.ts]: @P0lip Fixed

File: src/utils/pivotExceptions.ts (hunk 16 above)
Comment (nulltoken) [reply src/utils/pivotExceptions.ts]: @P0lip Fixed

File: src/runner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/runner.ts]: 🎉 It took me some time (as Windows and tty don't play well together) but I managed to find a way to honor exceptions in a `stdin` context.

@P0lip @philsturgeon Could you please take a look at the proposal?

File: src/rulesets/mergers/exceptions.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+const normalize = ($ref: string, rulesetUri?: string): string => {
+  const source = extractSourceFromRef($ref);
+
+  if (typeof source !== 'string') {
+    throw new InvalidUriError(buildInvalidUriErrorMessage($ref, rulesetUri, 'Missing source'));
+  }
+
+  if (rulesetUri === void 0 && source !== '<STDIN>' && !isAbsolute(source)) {
```
Comment (nulltoken) [inline src/rulesets/mergers/exceptions.ts]: @P0lip What would be a good file location to define and export that const `<STDIN>` marker and avoid magic strings littering the code?

File: src/runner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/runner.ts]: As a side note, that more or less compelled me to find a sustainable way to develop in a xplat way from a Windows computer. I'm going to open a series of tiny pull requests to make contributing from Windows a bit easier.

File: docs/getting-started/rulesets.md [hunk 18]
```diff
@@ -249,6 +249,51 @@ The example above will run all of the rules defined in the `spectral:oas` rulese
 
 - [Rules relevant to OpenAPI v2 and v3](../reference/openapi-rules.md)
 
+## Selectively silencing some results
+
+From time to time, you want to ignore some specific results without turning off
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]: Please can we try to pick a word and stick to it? Using both "silencing" and "ignoring" is ok, but then we talk about exceptions.

Can we either `ignore`, or `except`? I think personally think `ignore` is more clear, and in line with `.eslintignore`, `.gitignore`, etc.

File: docs/getting-started/rulesets.md [hunk 19]
```diff
@@ -249,6 +249,51 @@ The example above will run all of the rules defined in the `spectral:oas` rulese
...
+## Selectively silencing some results
+
+From time to time, you want to ignore some specific results without turning off 
+the rule entirely. This may happen, for instance, when working with legacy APIs.
+
+The ruleset can be extended to define such exceptions through the optional `except` property.
+
+`except` describes a map of locations (expressed as Json paths) and rules that should be ignored.
+
+Locations be can either described as relatively to the ruleset or absolute paths.
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]:
```suggestion
Locations be can either described as relative to the ruleset or absolute paths.
```

File: docs/getting-started/rulesets.md [hunk 20]
```diff
@@ -249,6 +249,51 @@ The example above will run all of the rules defined in the `spectral:oas` rulese
...
+$ cat /path/to/the/document | spectral lint -r /path/to/ruleset.yaml
+```
+
+...provided the ruleset is defined this way:
+
+```yaml
+extends: spectral:oas
+
+except:
+  "<STDIN>#":
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]: I don't think we need to mess around with this. I think putting something in the ruleset which demands knowledge of exactly how it's being called is going to be more confusion than its worth.

I would expect to see people writing `#/info`, or `models/awful.json` to make it shut up about issues in info, or a particularly awful model. If that info ignore just so happens to quieten up two different infos im not sure people will be that upset. 

Let's try and avoid overthinking things, and leave room for more improvements later if we get feedback that they're required. I don't think entrance files are the biggest concern for people, and if they are a more specific exception can be written, so this will possibly not come up for a long time - if ever. 🎲

Comment (nulltoken) [reply docs/getting-started/rulesets.md]: @philsturgeon Thanks for the detailed feedback!

The current implementation was designed to help make one of the initial requirements happen (from https://github.com/stoplightio/spectral/issues/747#issuecomment-568494226)

image

As such, the current syntax expects a file and a non ambiguous path within it that should be ignored.

Of course, the model can be extended to support more "handwritten" scenarios:

- `models/awful.json: [ rule1, rule2]`: Would ignore rule1 and rule2 results wheverever in the file
- `#/info": [rule1, rule2]`: Would ignore rule1 and rule2 results in any processed document provided they originate from a path that's `#/info` **or below (eg. `#/info/sub/path/1`)**. This would be in line with @P0lip's [comment](https://github.com/stoplightio/spectral/pull/939#discussion_r371174774)

Would that fit you better?

Comment (philsturgeon) [reply docs/getting-started/rulesets.md]: Ok if that’s the goal just ignore STDIN entirely and support files.

File: src/rulesets/mergers/exceptions.ts (hunk 17 above)
Comment (P0lip) [reply src/rulesets/mergers/exceptions.ts]: Good question.
I'd say `src/document.ts`

File: docs/getting-started/rulesets.md [hunk 21]
```diff
@@ -249,6 +249,51 @@ The example above will run all of the rules defined in the `spectral:oas` rulese
...
+    - oas3-api-servers
+  "<STDIN>#/info":
```
Comment (P0lip) [inline docs/getting-started/rulesets.md]: Hm, not a fan of it tbh.
How about `./#info`? This would be more universal IMHO, as you could also target the input document, i.e. it'd target `foo.yaml` in when the following command `spectral lint foo.yaml` is executed.

Comment (philsturgeon) [reply docs/getting-started/rulesets.md]: I've been recommending we just ditch STDIN support entirely it's not worth the hassle generally speaking we should plan features before we get people working on them :D 

This is trying to solve the Studio use case, and if people ask for this we can bring it in, but until then its scope creep.

File: src/rulesets/mergers/exceptions.ts (hunk 17 above)
Comment (nulltoken) [reply src/rulesets/mergers/exceptions.ts]: @P0lip Fixed

File: docs/getting-started/rulesets.md (hunk 18 above)
Comment (nulltoken) [reply docs/getting-started/rulesets.md]: @philsturgeon Fixed. Do you like it better?

File: docs/getting-started/rulesets.md (hunk 19 above)
Comment (nulltoken) [reply docs/getting-started/rulesets.md]: @philsturgeon Fixed

File: docs/getting-started/rulesets.md (hunk 20 above)
Comment (nulltoken) [reply docs/getting-started/rulesets.md]: @philsturgeon Dropped. STDIN and laxer syntax will be tracked through separate issues to keep that PR on scope.

File: src/rulesets/mergers/exceptions.ts [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    sourceRules.forEach(r => {
+      if (r.length === 0) {
+        throw new Error(buildErrorMessagePrefix(location, baseUri) + 'A rule with an empty name has been provided.');
+      }
+      set.add(r);
+    });
+
+    target[normalizedLocation] = [...set].sort();
```
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/rulesets/mergers/exceptions.ts]: @P0lip Fixed

File: src/rulesets/mergers/exceptions.ts [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+
+  const path = rulesetUri === undefined || isAbsolute(source) ? source : join(rulesetUri, '..', source);
+
+  return pathNormalize(path) + pointer;
+};
+
+const buildErrorMessagePrefix = ($ref: string, rulesetUri?: string): string => {
+  let prefix = '';
+
+  if (rulesetUri !== undefined) prefix += `in ruleset \`${rulesetUri}\`, `;
```
[the remark this answers is not in the document]
Comment (nulltoken) [reply src/rulesets/mergers/exceptions.ts]: @P0lip Fixed

File: docs/getting-started/rulesets.md (hunk 21 above)
Comment (nulltoken) [reply docs/getting-started/rulesets.md]: @P0lip @philsturgeon STDIN support has been dropped. A warning is now generated when the document is consumed from stdin and the ruleset contains exceptions.

Comment (P0lip) [reply docs/getting-started/rulesets.md]: > This is trying to solve the Studio use case, and if people ask for this we can bring it in, but until then its scope creep.

Studio should not be affected too much, since it doesn't use STDIN.

File: src/document.ts [hunk 24]
```diff
@@ -6,6 +6,8 @@ import { IParser } from './parsers/types';
 import { IRuleResult } from './types';
 import { startsWithProtocol } from './utils';
 
+export const StdIn: string = '<STDIN>';
```
Comment (P0lip) [inline src/document.ts]:
```suggestion
export const STDIN: string = '<STDIN>';
```

File: docs/getting-started/rulesets.md (hunk 18 above)
Comment (philsturgeon) [reply docs/getting-started/rulesets.md]: Yeah this is great! I'll give it a lick of paint after so you can focus on writing amazing code instead of mucking around with minor tweaks.

File: docs/getting-started/rulesets.md (hunk 21 above)
Comment (philsturgeon) [reply docs/getting-started/rulesets.md]: @P0lip that's what im trying to say. We are trying to get this working for the studio use case, which doesnt need STDIN, so we don't need to worry about STDIN here.

File: src/document.ts (hunk 24 above)
Comment (nulltoken) [reply src/document.ts]: @P0lip Fixed
